src/main/java/htsjdk/samtools/cram/structure/SubstitutionMatrix.java

```diff
@@ -164,44 +140,117 @@ public SubstitutionMatrix(final byte[] matrix) {
         baseByCode['N'][(encodedMatrixBytes[4] >> 2) & 3] = 'G';
         baseByCode['N'][(encodedMatrixBytes[4]) & 3] = 'T';
 
-        for (final byte refBase : SUBSTITUTION_BASES_UPPER) {
+        for (final SubstitutionBase refBase : SubstitutionBase.values()) {
             for (byte code = 0; code < CODES_PER_BASE; code++) {
-                codeByBase[refBase][baseByCode[refBase][code]] = code;
+                codeByBase[refBase.getBase()][baseByCode[refBase.getBase()][code]] = code;
             }
         }
     }
 
+    /**
+     * Given a reference base and a read base, find the corresponding substitution code
+     * @param refBase reference base being substituted
+     * @param readBase read base to substitute for the reference base
+     * @return code to be used for this refBase/readBase pair
+     */
+    public byte code(final byte refBase, final byte readBase) {
+        if (Character.isLowerCase((char) refBase)) {
+            log.warn(String.format("CRAM: Attempt to generate a substitution code for a lower case reference base '%c'", (char) refBase));
+        }
+        return codeByBase[refBase][readBase];
+    }
+
+    /**
+     * Given a reference base and a substitution code, return the corresponding substitution base.
+     * @param refBase reference base being substituted
+     * @param code substitution code
+     * @return base to be substituted for this (refBase, code) pair
+     */
+    public byte base(final byte refBase, final byte code) {
+        final byte base = baseByCode[refBase][code];
+        if (base == NO_BASE) {
+            // attempt to retrieve a code for a reference base that isn't in the substitution matrix
+            throw new IllegalArgumentException(String.format("CRAM: Attempt to retrieve a substitution base for '%c'", (char) refBase));
+        }
+        return base;
+    }
+
     /**
      * Return this substitution matrix as a byte array in a form suitable for serialization.
-     * @return
+     * @return the encoded matrix in serialized form
      */
     public byte[] getEncodedMatrix() {
         return encodedMatrixBytes;
     }
 
+    @Override
+    public String toString() {
+        final StringBuilder stringBuilder = new StringBuilder();
+        for (final SubstitutionBase r : SubstitutionBase.values()) {
+            stringBuilder.append((char) r.getBase());
+            stringBuilder.append(':');
+            for (int i = 0; i < CODES_PER_BASE; i++) {
+                stringBuilder.append((char) baseByCode[r.getBase()][i]);
+            }
+            stringBuilder.append('\t');
+        }
+        // lower case substitutions
+        for (final SubstitutionBase r : SubstitutionBase.values()) {
+            stringBuilder.append(Character.toLowerCase((char)r.getBase()));
+            stringBuilder.append(':');
+            for (int i = 0; i < CODES_PER_BASE; i++) {
+                stringBuilder.append((char) baseByCode[Character.toLowerCase((char)r.getBase())][i]);
```

Comment:
you can cache here the return from ```baseByCode[Character.toLowerCase((char)r.getBase())]```... in fact you could write this for as ```stringBuilder.append(new String(baseByCode[Character.toLowerCase((char)r.getBase())], 0, CODES_PER_BASE))``` if CODES_PER_BASE is the same length as the array you can skip the two last int arguments.

Reply:
That said I don't know if performance is better adding the bytes (casted into chars) one by one, is possible.

Reply:
I added the return value caching to improve readability, but otherwise perf is not really an issue for this method, which code is only used in debugging anyway.